InteractiveUtils: Fully support broadcasting expressions for code introspection macros

serenity4 opened the pull request:
This PR includes a fix and a feature for code introspection macros (`@code_typed`, `@code_llvm` and friends *but not* `@which`, `@edit`, etc):
- Fixes a bug for expressions of the form `f.(x; y = 3)`, for which keyword arguments were not properly handled and led to an internal error.
- Adds support for broadcasting assignments of the form `x .= f(y)`, `x .<<= f.(y, z)`, etc.

The way this was (and still is) implemented is by constructing a temporary function, `f(x1, x2, x3, ...) = <body>` and feeding that to code introspection functions. This trick doesn't apply to `@which` and `@edit`, which need to map to a single function call (we could arguably choose to target `materialize`/`materialize!`, but this behavior could be a bit surprising and difficult to support).

The switch differentiating the families of macro `@code_typed`/`@code_llvm` and `@which`/`@edit` etc was further exposed as an additional argument to `gen_call_with_extracted_types` and `gen_call_with_extracted_types_and_kwargs`, which default to the previous behavior (differentiating them based on whether their name starts with `code_`). The intent is to allow other macros such as `Cthulhu.@descend` to register themselves as code introspection macros. Quick tests indicate that it works as intended, e.g. with this PR Cthulhu supports `@descend [1, 2] .+= [2, 3]` (or equivalently, as added in #57909, `@descend ::Vector{Int} .+= ::Vector{Int}`).

I originally just went for the fix, and after some refactoring I realized the feature was very straightforward to implement.

serenity4 committed:
Properly handle broadcasting expressions in presence of kwargs

serenity4 committed:
Support broadcasting assignments

serenity4 committed:
Expose `is_code_macro` as an optional argument

serenity4 committed:
Add NEWS.md entry

serenity4 committed:
Merge branch 'master' of github.com:JuliaLang/julia into fix-broadcast-with-kwargs

serenity4 committed:
Turn `is_code_macro` into a keyword argument

serenity4 committed:
Merge branch 'master' of github.com:JuliaLang/julia into fix-broadcast-with-kwargs

topolarity commented on stdlib/InteractiveUtils/src/macros.jl:
```diff
@@ -318,7 +347,7 @@ Same behaviour as `gen_call_with_extracted_types` except that keyword arguments
 The keyword arguments must be given before the mandatory argument.
 """
-function gen_call_with_extracted_types_and_kwargs(__module__, fcn, ex0)
+function gen_call_with_extracted_types_and_kwargs(__module__, fcn, ex0; is_code_macro = default_is_code_macro(fcn))
```
I know you're just re-factoring the existing `codemacro` ugly-ness, but is it possible to name `is_code_macro` by its effect on the output (e.g. `wrap_broadcast_calls` ?), rather than the calling context?

serenity4 replied:
Yeah, I'm not a fan of `is_code_macro` but could not come up with anything better.

The current `is_code_macro` parameter also directs how we should handle `getproperty` calls for `@which`, `@edit`, etc. where if the first argument to `getproperty` is a module, then these are interested in the semantic reference rather than the `getproperty` call itself (well, currently only `@which` supports this, but with https://github.com/JuliaLang/julia/pull/53539 we'd be able to support most of the other non-code macros).

That one is a bit tricky, it might need a second parameter if we want to precisely name the differences in behavior across `code`/non-`code` macros in terms of their effects. Although, I guess the fundamental distinction is whether we are interested in a source location/reference (if so, wrapping anything as we do with broadcast calls destroys provenance, and we must disallow it), so we might as well rename `is_code_macro` to the negation of `reflect_into_source`/`reflect_into_reference` or something similar (if you have other ideas I'm all for it).

At least, I think if we introduce a parameter like `wrap_broadcast_calls` I feel like it should probably be named `allow_function_wrap` or similar to define the class of transformations we are allowed to apply, in case we need introduce a similar one. 

I'd be happy to have your thoughts on that.

topolarity replied:
Makes sense. I do think a name like `allow_function_wrap` or `preserve_provenance` might be a marginal improvement, even if it's not 100% comprehensive as a behavioral specification

I trust your judgement here in any case - I'd say pick your favorite for now (incl. the status quo if you want) and we can land that

serenity4 replied:
I decided to go with `is_source_reflection` in f02d397c1e7be034d24460c0924321b5239c39e1, and I also removed some special handling we had for `which` behind a more accessible `supports_binding_reflection` parameter. I'd love to get your feedback on that if you have any.

topolarity replied:
I'm a fan of the new changes.

It's still a bit context-sensitive (refers to the caller, rather than callee behavior), but at least it makes the diverging behavior clear and explains a bit of what they're tackling so clearly an improvement.

topolarity commented on stdlib/InteractiveUtils/src/macros.jl:
```diff
@@ -158,13 +187,15 @@ function merge_namedtuple_types(nt::Type{<:NamedTuple}, nts::Type{<:NamedTuple}.
     # assignments get bypassed: @edit a = f(x) <=> @edit f(x)
     if isa(ex0, Expr) && ex0.head == :(=) && isa(ex0.args[1], Symbol) && isempty(kws)
-        return gen_call_with_extracted_types(__module__, fcn, ex0.args[2])
+        return gen_call_with_extracted_types(__module__, fcn, ex0.args[2], Expr[]; is_code_macro)
```
```suggestion
        return gen_call_with_extracted_types(__module__, fcn, ex0.args[2], kws; is_code_macro)
```

Not a big deal, but saves an allocation

serenity4 committed:
Implement suggestion

serenity4 committed:
Merge branch 'master' of github.com:JuliaLang/julia into fix-broadcast-with-kwargs

serenity4 committed:
Add two comments

topolarity commented on stdlib/InteractiveUtils/test/runtests.jl:
```diff
@@ -619,7 +634,7 @@ end
     @test_throws err @code_lowered 1.0
 
-    @test_throws "is too complex" @code_lowered a .= 1 + 2
+    @test_throws "dot expressions are not lowered to a single function call" @which a .= 1 + 2
```
That's a much clearer error message, nice!

topolarity commented on stdlib/InteractiveUtils/test/runtests.jl:
```diff
@@ -505,6 +512,14 @@ a14637 = A14637(0)
+@test (@code_typed optimize=false [1, 2] .<<= 2)[2] == Vector{Int}
+@test (@code_typed optimize=false [1, 2.0] .= 1 .+ [2, 3])[2] == Vector{Float64}
+@test (@code_typed optimize=false [1, 2.0] .= 1 .+ round.(base = 1, [1, 3]; digits = 3))[2] == Vector{Float64}
+@test (@code_typed optimize=false [1] .+ [2])[2] == Vector{Int}
```
May be worth adding an argument-position broadcasting case like `v[5] = round.(data)`, where the broadcast isn't supposed to have any effect vs. the existing behavior

topolarity reviewed (approved):
The structure looks right to me, although I didn't do a close check for correct-ness

Thanks much, @serenity4 !

serenity4 committed:
Address a few review comments

serenity4 commented on stdlib/InteractiveUtils/src/macros.jl:
```diff
@@ -32,30 +32,59 @@ function get_typeof(@nospecialize ex)
-            push!(args, ex)
-            return newarg, i+1
+    if is_broadcasting_expr(ex)
+        (start, branches) = is_broadcasting_assignment(ex) ? (1, ex.args) : isexpr(ex, :.) ? (1, ex.args[2].args) : (2, ex.args)
```
I have to admit, that was ugly.

serenity4 committed:
Replace `is_code_macro` with `is_source_reflection`

Also includes another `supports_binding_reflection` parameter
to remove the special-cased support for `@which` reflection into
bindings.

serenity4 committed:
Update comment

topolarity commented:
CI is still green here, so I'm going to take the liberty to merge this 👍

serenity4 commented:
Thanks for the review @topolarity!
